### PR TITLE
Adding benchmarks, improving bench.py, creating graphs

### PR DESCRIPTION
**Pull request by msooseth**
Adds benchmarks, improves bench.py, generates graphs, add JSON and CSV outputs. Uses runlim for running and `/usr/bin/time` for timing.

### CODE REVIEW

File: foundry.toml [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+[fuzz]
+runs = 10
+max_test_rejects = 100
+seed = '0x4242'
+dictionary_weight = 40
+include_storage = true
+include_push_bytes = true
+
+# See more config options https://github.com/foundry-rs/foundry/tree/master/config
```
Comment (d-xo) [inline foundry.toml:14]: I think we can remove these?

**Review by d-xo — APPROVED** [review]
so nice 🤩🤩 just a few small comments & questions

File: notes.txt [hunk 2]
```diff
@@ -0,0 +1,8 @@
+TODO:
+* run `forge install foundry-rs/forge-std`
+
+Notes:
+* Need graphs via graphivz: CDF, pairwise comparison, etc.
+* Need sqlite
+* Need HTML generated
+* Beware, the timing is almost SURELY wrong. We need to run via `/usr/bin/time --verbose` and we need to be careful about multi-threading
```
[the remark this answers is not in the document]
Comment (msooseth) [reply notes.txt:8]: Haha, no, sorry, this was not meant to be committed :) Removing.

File: src/common/math.sol
[the remark this answers is not in the document]
Comment (msooseth) [reply src/common/math.sol:1]: This is used in `src/unsafe/ds-test/dsProveFail.sol`. I guess we could move to safe math, but I didn't want to change the test cases?

File: src/common/token.sol
[the remark this answers is not in the document]
Comment (msooseth) [reply src/common/token.sol:1]: Used by `src/unsafe/ds-test/dsProveFail.sol` again. Should I try to remove this? Sorry, I didn't want to change the test cases :S

[the remark this answers is not in the document]
Comment (msooseth) [reply src/common/token.sol:1]: Is this the right diff?

```
diff --git a/src/unsafe/ds-test/dsProveFail.sol b/src/unsafe/ds-test/dsProveFail.sol
index 9408bdb..61788ea 100644
--- a/src/unsafe/ds-test/dsProveFail.sol
+++ b/src/unsafe/ds-test/dsProveFail.sol
@@ -1,8 +1,7 @@
 pragma solidity ^0.8.19;
 
 import "ds-test/test.sol";
-import "src/common/token.sol";
-import "src/common/math.sol";
+import "src/common/erc20.sol";
 
 contract Withdraw {
     receive() external payable {}
@@ -13,17 +12,17 @@ contract Withdraw {
     }
 }
 
-contract SolidityTest is DSTest, DSMath {
-    DSToken token;
+contract SolidityTest is DSTest {
+    ERC20 token;
     Withdraw withdraw;
 
     function setUp() public {
-        token = new DSToken("TKN");
+        token = new ERC20("TKN", "T", 16);
         withdraw = new Withdraw();
     }
 
     function prove_transfer(uint supply, address usr, uint amt) public {
-        token.mint(supply);
+        token.mint(address(this), supply);
 
         uint prebal = token.balanceOf(usr);
         token.transfer(usr, amt);
```

Sorry, my understanding of this is quite low.

File: src/common/math.sol
[the remark this answers is not in the document]
Comment (msooseth) [reply src/common/math.sol:1]: OK, I think with the below PR I moved to safemath (default is solc now).

File: src/common/token.sol
[the remark this answers is not in the document]
Comment (d-xo) [reply src/common/token.sol:1]: looks good! Only thing I would change is to construct `ERC20` with 18 decimals:

```solidity
token = new ERC20("TKN", "T", 18);
```

[the remark this answers is not in the document]
Comment (msooseth) [reply src/common/token.sol:1]: Sorry, yes, remembered it wrongly, it's supposed to be 18, not 16. Fixed, thanks!

**Pull request merged**